Deprecate and move fabric-rendering-registries-v1 into fabric-rendering-v1

altrisi opened the pull request:
Resolves #1383.

Of course keeps all public API methods and classes functional, though just they call the new API instead.

Used a Java 9 feature (private methods in interfaces) in order to register a handler to invoke the old event from the new one.

altrisi committed:
Deprecate and move fabric-rendering-registries-v1 into fabric-rendering-v1

altrisi committed:
Make the event work again

altrisi committed:
Use lambda for RegistrationHelper

altrisi committed:
Checkstyle in rendererregistries

altrisi committed:
Checkstyle for rendering-v1

Technici4n commented on fabric-renderer-registries-v1/src/main/resources/fabric-renderer-registries-v1.mixins.json:
```diff
@@ -3,11 +3,6 @@
   "package": "net.fabricmc.fabric.mixin.client.renderer.registry",
```
You can drop this file entirely (and remove it from the mod .json as well)

altrisi replied:
Done.

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/BlockEntityRendererRegistry.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Helper class for registering BlockEntityRenderers.
+ */
+public interface BlockEntityRendererRegistry {
+	BlockEntityRendererRegistry INSTANCE = new BlockEntityRendererRegistryImpl();
+
+	/**
+	 * Register a BlockEntityRenderer for a BlockEntityType. Can be called clientside before the world is rendered.
+	 *
+	 * @param blockEntityType the {@link BlockEntityType} to register a renderer for
+	 * @param blockEntityRendererFactory a {@link class_5614} that creates a {@link BlockEntityRenderer}, called
+	 *                            when {@link BlockEntityRenderDispatcher} is initialized or immediately if the dispatcher
+	 *                            class is already loaded
+	 * @param <E> the {@link BlockEntity}
+	 */
+	<E extends BlockEntity> void register(BlockEntityType<E> blockEntityType, BlockEntityRendererFactory<? super E> blockEntityRendererFactory);
+}
```
I'd say this is a good opportunity to move to a static method instead of an `INSTANCE` field, with the implementation in a static method in `BlockEntityRendererRegistryImpl`.

altrisi replied:
Done, however, why couldn't it be inlined directly in the API class, like it's done in other classes?

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EntityRendererRegistry.java:
```diff
@@ -0,0 +1,43 @@
+
+import net.fabricmc.fabric.impl.client.rendering.EntityRendererRegistryImpl;
+
+/**
+ * Helper class for registering EntityRenderers.
+ */
+public interface EntityRendererRegistry {
+	EntityRendererRegistry INSTANCE = new EntityRendererRegistryImpl();
+
+	/**
+	 * Register a BlockEntityRenderer for a BlockEntityType. Can be called clientside before the world is rendered.
+	 *
+	 * @param entityType the {@link EntityType} to register a renderer for
+	 * @param entityRendererFactory a {@link EntityRendererFactory} that creates a {@link EntityRenderer}, called
+	 *                            when {@link EntityRenderDispatcher} is initialized or immediately if the dispatcher
+	 *                            class is already loaded
+	 * @param <E> the {@link Entity}
+	 */
+	<E extends Entity> void register(EntityType<? extends E> entityType, EntityRendererFactory<E> entityRendererFactory);
+}
```
Same comment here regarding moving away from `INSTANCE` fields in favor of static methods.

Technici4n commented:
Thanks for the PR! I noticed this is the chance to move away from the old `INSTANCE` pattern, if you don't mind.

altrisi committed:
Remove mixins file

And add two `forRemoval`

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/BlockEntityRendererRegistry.java:
```diff
@@ -0,0 +1,43 @@
+	 * Register a BlockEntityRenderer for a BlockEntityType. Can be called clientside before the world is rendered.
+	 *
+	 * @param blockEntityType the {@link BlockEntityType} to register a renderer for
+	 * @param blockEntityRendererFactory a {@link class_5614} that creates a {@link BlockEntityRenderer}, called
```
mind update the jd here to refer to the right class?

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EntityModelLayerRegistry.java:
```diff
@@ -0,0 +1,71 @@
+ * @deprecated Experimental feature, may be removed or changed without further notice: Snapshot feature.
+ */
+@Deprecated
+@ApiStatus.Experimental
```
can we raise this out of the exp state?

Technici4n replied:
Most likely yes. I'm not aware of any issue with this addition, and it should be stable to have for 1.17 now.

altrisi committed:
Raise EntityModelLayerRegistry out of experimental

And update javadoc

liach commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EntityModelLayerRegistry.java:
```diff
@@ -58,8 +52,6 @@ private EntityModelLayerRegistry() {
 
 	@FunctionalInterface
-	@ApiStatus.Experimental
-	@Environment(EnvType.CLIENT)
```
uhh, you removed an annotation incorrectly

altrisi committed:
Change INSTANCE to a static method in two registries

altrisi committed:
Readd unwillingly removed annotation

altrisi commented:
I've made the requested changes: Took EntityModelLayerRegistry out of experimental API and changed the `INSTANCE` pattern to a static method calling another static method in the Impl class.

My question is: Can't that be inlined and have the implementation directly in the API class, given it's a static method in a final class anyway? Other classes in the API already do that, including EntityModelLayerRegistry, that comes from the same module.

Technici4n commented:
Thanks!

I think inlining is fine if the internal logic is small enough that it doesn't hinder readability of the API class, which should be the case here. To be confirmed by @sfPlayer1.

liach commented:
i think you can just do that like what the jdk does.

altrisi committed:
Disallow constructing API instances

altrisi commented:
Actually I don't think it's worth it, given the impl classes still have to exist to hide the setup methods and fields that mixins reference.

Will keep it like this unless you think it'd still make sense.

liach commented:
Sure, feel free to choose whatever impl as long as it is not noted to be inefficient and exposes only our desired contents.

altrisi commented:
(then this should be ready for reviews, triage request got removed after annotation comment)

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/LivingEntityFeatureRendererRegistrationCallback.java:
```diff
@@ -0,0 +1,80 @@
+	 *
+	 * <p>This is not meant for implementation by users of the API.
+	 */
+	interface RegistrationHelper {
```
You can add `@ApiStatus.NonExtendable` for this.

altrisi replied:
Done.

Technici4n commented on fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EntityModelLayerRegistry.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * A helpers for registering entity model layers and providers for the layer's textured model data.
+ */
+@Environment(EnvType.CLIENT)
```
This one is `@Environment(EnvType.CLIENT)` and not the others. Should all of them have it?

altrisi replied:
Added them, makes sense since they all reference classes only present in the client.

Technici4n commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/EntityModelLayerRegistry.java:
```diff
@@ -16,25 +16,18 @@
  */
-@Deprecated
-@ApiStatus.Experimental
+@Deprecated(forRemoval = true)
```
Not sure about the use of `(forRemoval = true)`, as we don't have plans to remove it for the moment... it should not be a maintenance burden.

liach replied:
this should be, as jdk now calls forremoval as terminally deprecated

Technici4n replied:
Hmmm, the documentation says that
> `forRemoval=true` indicates that the API is subject to removal in a future release.

So it should be fine to have them. Can be marked as resolved.

altrisi replied:
I assumed this was the case since the issue was about reducing the number of modules, and they can't be reduced if it's not removed at some point in the future.

Technici4n reviewed (commented):
A few minor inquiries, otherwise this PR seems appropriate. I'll do a bit of local testing when I find the time this week.

altrisi committed:
Added Environment and ApiStatus annotations

altrisi reviewed (commented):
Made the requested changes.

(edit: wow github messed up the replies to the comments in here, see above for the actual conversations)

Technici4n commented on fabric-rendering-v1/src/testmod/resources/fabric.mod.json:
```diff
@@ -10,7 +10,8 @@
     "client": [
-      "net.fabricmc.fabric.test.rendering.client.WorldRenderEventsTests"
+      "net.fabricmc.fabric.test.rendering.client.WorldRenderEventsTests",
+      "net.fabricmc.fabric.test.renderer.registry.FeatureRendererTest"
```
```suggestion
      "net.fabricmc.fabric.test.rendering.client.FeatureRendererTest"
```

Technici4n reviewed (approved):
Looks fine, and I confirm that the living entity feature renderer registration callback works with that small adjustment. Haven't been able to test the others though.

altrisi committed:
Update testmod's entrypoint

Co-authored-by: Technici4n <13494793+Technici4n@users.noreply.github.com>

altrisi commented:
Bump, is there anything missing?

Player3324 commented:
I'll have a closer look in a bit, `@Deprecated(forRemoval = true)` should be just `@Deprecated` since we technically don't remove the module but leave it as legacy

Player3324 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/EntityRendererRegistry.java:
```diff
@@ -26,7 +26,10 @@
  */
+@Deprecated(forRemoval = true)
 public interface EntityRendererRegistry {
 	EntityRendererRegistry INSTANCE = new EntityRendererRegistryImpl();
```
may as well define this inline (anonymous class?), same in BlockEntityRendererRegistry

Player3324 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/LivingEntityFeatureRendererRegistrationCallback.java:
```diff
@@ -63,11 +62,26 @@
 	void registerRenderers(EntityType<? extends LivingEntity> entityType, LivingEntityRenderer<?, ?> entityRenderer, RegistrationHelper registrationHelper, EntityRendererFactory.Context context);
 
+	private static Event<LivingEntityFeatureRendererRegistrationCallback> createEvent() {
+		var event = EventFactory.createArrayBacked(LivingEntityFeatureRendererRegistrationCallback.class, callbacks -> (entityType, entityRenderer, registrationHelper, context) -> {
```
no `var` use please

Player3324 reviewed (commented):
Mostly good, thanks and sorry for the delay. Besides the deprecation change request from my above comment there are 2 minor details inline and I think the test mod in the old module should go, it only conflicts with the "new" copy.

altrisi committed:
Remove renderer-registries testmod

altrisi committed:
Remove `var` usage

altrisi committed:
Don't specify `forRemoval = true` in `@Deprecated`

altrisi committed:
Inline old registry implementations in anonymous classes

altrisi commented:
Done: Removed old testmod, changed `var` and inlined the old registry implementations.

Also removed the `forRemoval = true` from `@Deprecated`, though I thought they would fit since the issue mentioned "reduce the number of modules in Fabric API", which isn't really possible without removing it at some point. There was also a bit of conversation about it in https://github.com/FabricMC/fabric/pull/1470#discussion_r650574661.

Player3324 commented:
Thanks, it's effectively still reduced by becoming a legacy module. We'll change how they are exposed in the near future, so they don't bother devs anymore while still retaining compat for end users.

altrisi committed:
Merge branch '1.17' into deprecate-rendering-registries

altrisi commented:
Fixed conflicts with recent additions to rendering-v1.

liach commented:
@modmuss50 Mind check it out?

Juuxel commented on fabric-rendering-v1/src/main/resources/fabric-rendering-v1.mixins.json:
```diff
@@ -8,7 +8,12 @@
+    "EntityModelsMixin",
+    "LivingEntityRendererAccessor",
+    "MixinBlockEntityRenderers",
+    "MixinEntityRenderers"
```
Preferably this would be sorted alphabetically like in other modules, but this is fine too.

Juuxel reviewed (approved):
Nice! This also will resolve #427 if the deprecated module is not kept around forever (which'll probably be the case...).
